evm/vm/forks/sharding/vmc_handler.py

```diff
@@ -118,44 +103,79 @@ def _get_new_logs(self, shard_id):
         for log in new_logs:
             yield parse_collation_added_data(log['data'])
 
-    def get_next_log(self, shard_id):
-        new_logs = self._get_new_logs(shard_id)
-        self.new_logs[shard_id].extend(new_logs)
-        if len(self.new_logs[shard_id]) == 0:
+    def get_next_log(self):
+        new_logs = self._get_new_logs()
+        self.new_logs.extend(new_logs)
+        if len(self.new_logs) == 0:
             raise NextLogUnavailable("No more next logs")
-        return self.new_logs[shard_id].pop()
+        return self.new_logs.pop()
 
     # TODO: this method may return wrong result when new logs arrive before the logs inside
     #       `self.new_logs` are consumed entirely. This issue can be resolved by saving the
     #       status of `new_logs`, `unchecked_logs`, and `current_score`, when it start to run
     #       `GUESS_HEAD`. If there is a new block arriving, just restore them to the saved status,
     #       append new logs to `new_logs`, and re-run `GUESS_HEAD`
-    def fetch_candidate_head(self, shard_id):
+    def fetch_candidate_head(self):
         # Try to return a log that has the score that we are checking for,
         # checking in order of oldest to most recent.
         unchecked_logs = pipe(
-            self.unchecked_logs[shard_id],
+            self.unchecked_logs,
             enumerate,
             tuple,
             reversed,
             tuple,
         )
-        current_score = self.current_score[shard_id]
+        current_score = self.current_score
 
         for idx, logs_entry in unchecked_logs:
             if logs_entry['score'] == current_score:
-                return self.unchecked_logs[shard_id].pop(idx)
+                return self.unchecked_logs.pop(idx)
         # If no further recorded but unchecked logs exist, go to the next
         # is_new_head = true log
         while True:
             # TODO: currently just raise when there is no log anymore
-            self.unchecked_logs[shard_id].append(self.get_next_log(shard_id))
-            if self.unchecked_logs[shard_id][-1]['is_new_head'] is True:
+            self.unchecked_logs.append(self.get_next_log())
+            if self.unchecked_logs[-1]['is_new_head'] is True:
```

Comment:
If i'm reading this logic right.

- retrieve the new log with `get_next_log()`
- append to `self.unchecked_logs`
- if newest log `is_new_head == True`
    - break from loop and pop off newest log.

I think this is simpler written as:

```python
while True:
    log_entry = self.get_new_log()
    if log_entry['is_new_head']:
        break
    else:
        self.unchecked_logs.append(log_entry)

...
```

It avoids appending and then popping by only appending in the case that the new log isn't the new head.

Reply:
you're right. Thank you for pointing it out.
I've modified the code.